design-docs/handling-encodings.md

```diff
@@ -0,0 +1,87 @@
+archives (zip files, tar files, etc.)
+
+## Use Cases
+
+### File Content Encoding
+
+Any time we need to generate a text file, or to read, transform and write the contents of a text file, we need to know what 
+character encoding(s) to use.
+
+Common situations include:
+
+* Copying a file from one location to another.
+* Creating an archive (zip, tar, etc.).
+* Transforming text in a file.
+
+[GRADLE-1267](https://issues.gradle.org/browse/GRADLE-1267) describes just one instance of where this functionality is needed.
+
+### File Name Encoding
+
+Each file system has an idea of what character sets are used to represent files in the filesystem, and it's important to be 
+able to use filenames in archives that are encoded using the encoding of the target file system, which is not necessarily the same
+as the file system where the archive is created.
+
+This use case is why the Zip task and the War task have an `encoding` option added in 
+[Pull Request 499](https://github.com/gradle/gradle/pull/499) which allows creating a zip archive which contains files whose names 
+are encoded with a character encoding different from the default platform encoding.
+
+## Solution
+
+This design generally proposes that any task which needs to read and write text or encode file names in an archive, should support two configurable settings:
+
+1. `contentEncoding` - The encoding to be used to convert bytes to characters (when reading), and characters to bytes (when writing).
+2. `fileNameEncoding` - The encoding to be used for names of files in an archive.
+
+Both properties default to platform default encoding if they are not specified in the task's configuration block.
+
+Notes:
+
+ - reading and writing characters isn't limited to files. It also applies to archive entries, network communication, etc.
+ - some tasks might need to use an encoding to write that is different from the encoding used to read. For the most frequent case,
+   i.e. copying and transforming files, it's a reasonable assumption to expect the developer to choose the target encoding for the
+   source files, and thus to use a single `contentEncoding` both for reading and writing. By the way, when not filtering the 
+   files, the Copy task treats them as binary files, and thus uses the same encoding for the source and target files.
+
+## Implementation Plan
+
+The CopySpec interface and all its implementations should support a `contentEncoding` option, used only when the spec filters 
+the content (otherwise, the bytes of the source files are copied as is to the destination). This allows filtering the content 
+correctly from and to the file system, as well as from and to archives. 
+
+For the Zip, War, Ear and Tar tasks, we should add support a `fileNameEncoding` encoding option, that should thus be set on their
+parent AbstractArchiveTask.
+
+To be easier to specify in the build file, these properties should be of type `String`, and not of type `Charset`.
+
+## Testing Plan
```

Comment:
Another suggestion for testing is actually to ensure that no code "accidentally" uses the default encoding. It should always be specified (even if it is read from the system and that value is used, but explicitly). The way we ensure this consistency in Lucene and Elasticsearch is to use the [forbidden apis](https://github.com/policeman-tools/forbidden-apis) tool. You can see an example signatures file [here](https://github.com/apache/lucene-solr/blob/master/lucene/tools/forbiddenApis/lucene.txt#L20).

/cc @uschindler who might have other thoughts on this doc